### PR TITLE
network: introduce Family= setting in [NextHop] section

### LINKED ISSUE #18446: networkd: NextHop should not require Gateway= directive
The `[NextHop]` Section of systemd.network configuration requires a `Gateway=` directive.

Point-to-point link may not have a gateway address and assumes the peer on the other end is the gateway.

`Gateway=` directive should not be mandatory. And a `Family=` directive would be required when `Gateway=` is missing.

### PR DESCRIPTION
**Pull request by yuwata**
Closes #18446.

Split out of #18589.

### CODE REVIEW

File: src/network/networkd-nexthop.c [hunk 1]
```diff
@@ -538,3 +538,69 @@ int config_parse_nexthop_gateway(
...
+        assert(lvalue);
+        assert(rvalue);
+        assert(data);
+
+        r = nexthop_new_static(network, filename, section_line, &n);
+        if (r < 0)
+                return log_oom();
+
+        if (isempty(rvalue) &&
+            in_addr_is_null(n->family, &n->gw) != 0) {
```
Comment (keszybz) [inline src/network/networkd-nexthop.c:585]: Is the condition reverted here (== 0 instead would match the comment)?

File: src/network/networkd-nexthop.c [hunk 2]
```diff
@@ -483,13 +486,25 @@ int config_parse_nexthop_id(
...
+
+        r = safe_atou32(rvalue, &id);
         if (r < 0) {
                 log_syntax(unit, LOG_WARNING, filename, line, r,
                            "Could not parse nexthop id \"%s\", ignoring assignment: %m", rvalue);
                 return 0;
         }
+        if (id == 0) {
+                log_syntax(unit, LOG_WARNING, filename, line, 0,
+                           "Invalid nexthop id \"%s\", ignoring assignment: %m", rvalue);
```
Comment (keszybz) [inline src/network/networkd-nexthop.c:503]: I assume we don't care about backwards compat here. It's a fringe case.

File: man/systemd.network.xml [hunk 3]
```diff
@@ -1315,16 +1315,26 @@ IPv6Token=prefixstable:2002:da8:1::</programlisting></para>
       hops.</para>
 
       <variablelist class='network-directives'>
+        <varlistentry>
+          <term><varname>Id=</varname></term>
+          <listitem>
+            <para>The id of the nexthop. Takes an unsigned integer in the range 1…4294967295. If
```
Comment (keszybz) [inline man/systemd.network.xml]: "next hop" ?
"If unspecifed" → "If left unspecified"

**Review by keszybz — COMMENTED** [review]
I have trouble with the _is_null() functions. I think we could at some point do a mass replacement of
`in_addr_is_null() != 0` with `in_addr_is_set() == 0`. I think that'd be easier to reason about.

**Comment by poettering** [comment]
from my side all looks good

File: src/network/networkd-nexthop.c (hunk 1 above)
Comment (yuwata) [reply src/network/networkd-nexthop.c:585]: The condition is OK. The comment is updated to make it match the condition.

**Comment by yuwata** [comment]
@keszybz and @poettering Thank you for your review. All requests except for introducing `in_addr_is_set()` are addressed. The idea `in_addr_is_set()` is nice, but I'd like to leave that to another PR.

**Pull request merged**